### PR TITLE
Query History: Prevent viewers from accessing

### LINKED ISSUE #88722: Query History: Do not return results from API if user is in the viewer role
As a response to https://community.grafana.com/t/should-viewer-have-access-to-query-history-in-explore-menu/123023

This is not a critical security issue, but something to follow up on.

Results for Query History are returned via a GET to `api/query-history` - results are automatically filtered to be for the user who sent the request. 

In Grafana's permission system, the "viewer" role is generally limited from accessing query information. They can see dashboards, but cannot edit them to see the queries generating the results, nor can they access Explore. 

In the following scenario, a user in the viewer role would be able to access query data via query history:
1. The user was formerly not in a viewer role, and ran queries in Explore
2. The user noted the API call to get query history data
3. The user was then changed to be viewer
4. The user ran the API call before the query history cleanup was triggered (defaulting to 2 weeks), or they starred queries.

Actual Result: The user would see query data. They would only be able to access queries they had previously ran.
Expected Result: The API call to get query history data should not return results if the user is in the viewer role.

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

This adds a permissions check to ensure a user attempting to access query history has access to do so.

**Why do we need this feature?**

There is an edge case where a user had, then lost access, but still can access their previously ran queries. Viewers should not have access to queries. In Query History, users can only access their own queries, so this is a low priority security issue.

**Which issue(s) does this PR fix?**:

Fixes #88722

**Special notes for your reviewer:**
Tested this with an editor, viewer, and viewer with 
```
[users]
viewers_can_edit = true
```

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/queryhistory/api.go [hunk 1]
```diff
@@ -61,6 +62,10 @@ func (s *QueryHistoryService) createHandler(c *contextmodel.ReqContext) response
 // 401: unauthorisedError
 // 500: internalServerError
 func (s *QueryHistoryService) searchHandler(c *contextmodel.ReqContext) response.Response {
+	if (c.GetOrgRole() == org.RoleViewer && !s.Cfg.ViewersCanEdit) {
```
Comment (gelicia) [inline pkg/services/queryhistory/api.go:65]: Wanted to note I attempted to fix this by attaching the appropriate middleware permissions in https://github.com/grafana/grafana/blob/main/pkg/services/queryhistory/api.go#L18, but the role is strict and does not consider the viewers_can_edit config.

**Comment by ifrost** [comment]
I think we should also apply it to all query history endpoints 🤔

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Review by aocenas — APPROVED** [review]

**Comment by gelicia** [comment]
@ifrost I'm really sorry, I spaced on this one. I was asking the access team if they had anything for an exception for viewers_can_access and the outcome was a no. I thought about making a global auth middleware setting for this, but in the end I think it's a fairly unique scenario where trying to bake it into something core is overkill. I just added the same logic to every handler. 

Why I think this is unique: Viewers can edit rides this line where people can run queries, but not save anything. Explore access is fully allowed, dashboard panels can be edited, but not saved. Query history a.) is an outcome of running queries (good!) b.) saves something to the database (bad!) so its kind of in-between. I do think we just want to save stuff to the database for viewers here, and if it ever becomes an issue we can change the frontend logic to only save query history data to local storage for viewers.

**Comment by ifrost** [comment]
> @ifrost I'm really sorry, I spaced on this one. I was asking the access team if they had anything for an exception for viewers_can_access and the outcome was a no.

I feel like this should be implemented the same way as the access to Explore, i.e. the access being elevated when `viewers_can_edit = true`

* Explore uses [ac.ActionDatasourcesExplore](https://github.com/grafana/grafana/blob/62494248e3c26495f85fb049a5094dd17f70d7eb/pkg/api/api.go#L176), 
* which is additionally granted to the Viewer [here](https://github.com/grafana/grafana/blob/0aebb9ee3929f94157ebaaaa95776aa838b571db/pkg/api/accesscontrol.go#L73) 
* and it's using [`fixed:datasources:explorer` role](https://github.com/grafana/grafana/blob/0aebb9ee3929f94157ebaaaa95776aa838b571db/pkg/api/accesscontrol.go#L58)

We can have a new permission for QueryHistory specifically that is available only to Editors (or Viewers when viewers_can_edit is enabled). Otherwise we end up in two ways of handling viewers_can_edit which will be hard to maintain.

But anyway - Dashboards to their own magic when handling viewer_can_edit anyway, so I'm happy to merge it as it as as well and create a follow-up ticket to refactor it 👍

**Review by ifrost — APPROVED** [review]

**Pull request merged**